topics/atomic-slot-migration.md

```diff
@@ -0,0 +1,258 @@
+source node to a target node. Valkey 9.0 introduces a new option for migrating
+hash slots known as Atomic Slot Migration.
+
+## Background: The Existing Migration Process
+
+Prior to Valkey 9.0, a slot migration from a source node to a target node was
+performed through the following steps:
+
+1. Send `<target>` `CLUSTER SETSLOT <slot> IMPORTING <source>`
+2. Send `<source>` `CLUSTER SETSLOT <slot> MIGRATING <target>`
+3. Send `<source>` `CLUSTER GETKEYSINSLOT <slot> <count>`
+4. Send `<source>` `MIGRATE ...` for each key in the result of step 3
+5. Repeat 3 & 4 until no keys are left in the slot on `<source>`
+6. Send `<target>` `CLUSTER SETSLOT <slot> NODE <target>`
+7. Send `<source>` `CLUSTER SETSLOT <slot> NODE <target>`
+
+This was subject to the following problems:
+
+- **Higher latency for client operations**: All client writes and reads to keys
+  in the migrating hash slot were subject to `ASK` redirections, requiring
+  re-execution of the command on the target node, effectively doubling client
+  observed latency during the migration.
+- **Multi-key operation unavailability**: Since there was no guarantee all keys
+  in a multi-key operation were present on a single node, multi-key operations
+  like [`MSET`](../commands/mset.md) or [`MGET`](../commands/mget.md) were
+  subject to `TRYAGAIN` errors.
+- **Problems with large keys/collections**: Since the migration was performed
+  one key at a time, large keys (like a hash with many keys or a list with many
+  elements) needed to be sent as a single command. Serialization of a large key
+  required a large contiguous memory chunk on the source node and import of that
+  payload required a large CPU burst on the target node. In some cases, the
+  memory consumption is enough to trigger out-of-memory conditions on the source
+  node, and the CPU burst is large enough to cause a failover on the target
+  shard due to health probes not being served.
+- **Slot migration latency**: The overall latency of the slot migration was
+  bounded by how quickly the operator could send the `CLUSTER GETKEYSINSLOT` and
+  `MIGRATE` commands. Each batch of keys required a full round-trip-time between
+  the operator's machine and the cluster, leaving a lot of waiting time that
+  could be used to do data migration.
+- **Lack of resilience to failure**: If a failure condition was encountered, for
+  example, the hash slot will not fit on the target node, undoing the slot
+  migration is not well supported and requires replaying the listed steps in
+  reverse. In some cases, the hash slot may have grown while the migration was
+  underway, and may not fit on either the source or target node.
+
+## Performing an Atomic Slot Migration using `CLUSTER MIGRATESLOTS`
+
+Valkey 9.0 does not get rid of the existing slot migration option, but it does
+introduce Atomic Slot Migration as a second option. To perform an Atomic Slot
+Migration, an operator performs the following steps:
+
+1. Send `<source>`
```

Comment:
I think more of this context should move to `cluster migrateslots`, since that is the main command that end users will use to trigger this.